src/safeposix/syscalls/fs_calls.rs

```diff
@@ -910,6 +909,421 @@ impl Cage {
+                Stream(_) => {return 0;},
+                Socket(_) => {
+                    Self::_cleanup_socket(self, &fd, false);
+                 },
+                Pipe(pipe_filedesc_obj) => {
+                    let pipenumber = pipe_filedesc_obj.pipe;
+                    let read_references = Self::_lookup_refs_by_pipe_end(self, pipenumber, O_RDONLY);
+                    let write_references = Self::_lookup_refs_by_pipe_end(self, pipenumber, O_WRONLY);
+
+                    if write_references == 1 && pipe_filedesc_obj.flags == O_WRONLY {
+                        // let pipetable.pipenumber.eof = true;
+                    }
+
+                    if read_references + write_references == 1 {
+                        //del pipetable.pipenumber
+                    }
+                },
+                //TO DO: check IS_EPOLL_FD and if true, call epoll_object_deallocator (if necessary?)
+                File(normalfile_filedesc_obj) => {
+                    let inodenum = normalfile_filedesc_obj.inode;
+                    let inodeobj = mutmetadata.inodetable.get_mut(&inodenum).unwrap();
+                    let mut fobjtable = FILEOBJECTTABLE.write().unwrap();
+
+                    match inodeobj {
+                        Inode::File(ref mut normalfile_inode_obj) => {
+                            normalfile_inode_obj.refcount -= 1;
+
+                            //if there are still references to the file, then do nothing to the file
+                            if normalfile_inode_obj.refcount != 0 {
+                                return 0;
+                            }
+                            if fobjtable.contains_key(&inodenum) {
+                                //close and delete the file in the object table
+                                fobjtable.get(&inodenum).unwrap().close().unwrap();
+                                fobjtable.remove(&inodenum);
+                            }
+                            if normalfile_inode_obj.refcount == 0 {
+                                if normalfile_inode_obj.linkcount == 0  {
+                                    
+                                    //removing the file from the entire filesystem
+                                    mutmetadata.inodetable.remove(&inodenum);
+                                    let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
+                                    interface::removefile(sysfilename).unwrap();
+            
+                                } 
+                            }
+                        },
+                        Inode::Dir(ref mut dir_inode_obj) => {
+                            dir_inode_obj.refcount -= 1;
+
+                            //if there are still references to the file, then do nothing to the file
+                            if dir_inode_obj.refcount != 0 {
+                                return 0;
+                            }
```

Comment:
You still want to go through with removing the file descriptor from the fdtable even if the refcount is non-zero, the code will need to be refactored to allow that